### PR TITLE
feat: impl `increase` and `irate`/`idelta` in PromQL

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implement PromQL range functions `increase`, `idelta` and `irate`

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/promql/src/functions/idelta.rs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        assert_eq!(input.len(), 2);
+        let ts_array = extract_array(&input[0])?;
+        let value_array = extract_array(&input[1])?;
+
+        let ts_range: RangeArray = RangeArray::try_new(ts_array.data().clone().into())?;
+        let value_range: RangeArray = RangeArray::try_new(value_array.data().clone().into())?;
+        error::ensure(
+            ts_range.len() == value_range.len(),
+            DataFusionError::Execution(format!(
+                "input arrays should have the same length, found {} and {}",
```
Comment (killme2008) [inline src/promql/src/functions/idelta.rs:96]: Adds the function name into the error message?

File: src/promql/src/functions/idelta.rs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                ts_range.len(),
+                value_range.len()
+            )),
+        )?;
+        error::ensure(
+            ts_range.value_type() == DataType::Int64,
+            DataFusionError::Execution(format!(
+                "expect Int64 as time index array's type, found {} ",
```
Comment (killme2008) [inline src/promql/src/functions/idelta.rs:104]: The same with above.

File: src/promql/src/functions/idelta.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+            }
+
+            // if is delta
+            if !IS_RATE {
+                result_array.push(values[len - 1] - values[len - 2]);
+                continue;
+            }
+
+            // else is rate
+            let sampled_interval = (timestamps[len - 1] - timestamps[len - 2]) / 1000;
```
Comment (killme2008) [inline src/promql/src/functions/idelta.rs:155]: Can we add a comment here? Do we consume the timestamp unit is milliseconds here?

Comment (waynexia) [reply src/promql/src/functions/idelta.rs:143]: I add a todo comment. I've not figured out how to adapt other timestamp precisions to those functions. Plan to investigate this later after more functions are implemented.

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
@evenyag @yuanbohan  PTAL

File: src/promql/src/functions/increase.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            (3, 0),
+            (4, 3),
+            (5, 1),
+            (6, 0),
+            (7, 2),
+        ];
+        let range_array = RangeArray::from_ranges(values_array, ranges).unwrap();
+        increase_runner(range_array, vec![0.0, 0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 1.0]);
+    }
+}
```
Comment (yuanbohan) [inline src/promql/src/functions/increase.rs:168]: > [Breaks in monotonicity (such as counter resets due to target restarts) are automatically adjusted for.](https://prometheus.io/docs/prometheus/latest/querying/functions/#increase)

Is this feature covered?

Comment (waynexia) [reply src/promql/src/functions/increase.rs:188]: nice catch. I add `counter_reset()` to cover it

**Review by yuanbohan — APPROVED** [review]

**Pull request merged**